src/histogram.c

```diff
@@ -0,0 +1,236 @@
+			if (lbs[0] != 0) {
+				elems_edit[0] = (Datum) 0;
+				for (int j = 1; j < n; j++) {
+					elems_edit[j] = elems[j - 1];
+				}
+			}
+			else {
+				for (int j = 0; j < n; j++) {
+					elems_edit[j] = elems[j];
+				}
+			}
+			elems_edit[bucket] = (Datum) 0;
+			elems = elems_edit;
+		}
+
+		/* Set number of buckets */
+		dims[0] = n;
+	}
+
+	/* Increment the proper bucket */
+	elems[bucket-s] = elems[bucket-s] + (Datum) 1; 
+
+	/* Construct the state array */
+ 	state = construct_md_array(elems + lbs[0] - s, NULL, 1, dims, lbs, INT4OID, 4, true, 'i'); 
+
+	/* Return the integer array */
+	PG_RETURN_ARRAYTYPE_P(state); 
+}
+
+/* hist_combinefunc(ArrayType, ArrayType) => ArrayType */
+Datum
+hist_combinefunc(PG_FUNCTION_ARGS)
+{
+	MemoryContext 	aggcontext;
+	ArrayType 	*state1 = PG_ARGISNULL(0) ? NULL : PG_GETARG_ARRAYTYPE_P(0);
+	ArrayType 	*state2 = PG_ARGISNULL(1) ? NULL : PG_GETARG_ARRAYTYPE_P(1);
+
+	if (!AggCheckCallContext(fcinfo, &aggcontext))
+	{
+		/* Cannot be called directly because of internal-type argument */
+		elog(ERROR, "hist_combinefunc called in non-aggregate context");
+	}
+
+	if (state2 == NULL)
+		PG_RETURN_ARRAYTYPE_P(state1);
+
+	else if (state1 == NULL)
+		PG_RETURN_ARRAYTYPE_P(state2);
+
+	else 
+	{
+		Datum 	*s1, *s2, *result; // Datum array representations of state1, state2, and result
+
+		int     dims[1]; // 1-D array containing number of buckets used to construct result
+ 		int     lbs[1]; // 1-D array containing the lower bound used to construct result
+		int 	ubs; // upper bound used to construct histogram
+
+		/* Lower and upper bounds for state1 and state2 */
+		int 	lb1 = DirectFunctionCall2(array_lower, PointerGetDatum(state1), 1);
+		int 	lb2 = DirectFunctionCall2(array_lower, PointerGetDatum(state2), 1);
+		int 	ub1 = DirectFunctionCall2(array_upper, PointerGetDatum(state1), 1);
+		int 	ub2 = DirectFunctionCall2(array_upper, PointerGetDatum(state2), 1); 
+
+		/* State variables */
+		Oid    	i_eltype;
+	    int16  	i_typlen;
+	    bool   	i_typbyval;
+	    char   	i_typalign;
+	    int 	n;
+
+		/* Get bound extremities */ 
+		lbs[0] = (lb1 <= lb2) ? lb1 : lb2;
+		ubs = (ub1 >= ub2) ? ub1 : ub2;
+
+		dims[0] = ubs - lbs[0] + 1;
+		result = (Datum *) MemoryContextAlloc(aggcontext, sizeof(Datum) * (dims[0]));
+
+		/* Get state1 array element type */
+		i_eltype = ARR_ELEMTYPE(state1);
+		get_typlenbyvalalign(i_eltype, &i_typlen, &i_typbyval, &i_typalign);
+
+		/* Deconstruct state1 into s1 */
+		deconstruct_array(state1, i_eltype, i_typlen, i_typbyval, i_typalign, &s1, NULL, &n);
+
+		/* Get state2 array element type */
+		i_eltype = ARR_ELEMTYPE(state2);
+		get_typlenbyvalalign(i_eltype, &i_typlen, &i_typbyval, &i_typalign);
+
+		/* Deconstruct state2 into s2 */
+		deconstruct_array(state2, i_eltype, i_typlen, i_typbyval, i_typalign, &s2, NULL, &n); 
+
+		/* Initialize result array (which is zero-indexed in C) with zeroes */
+		for (int i = 0; i < dims[0] + lbs[0]; i++) {
```

Comment:
Again, I am wondering if it is not possible to just add the values of array2 to array1, avoiding copying and deconstruct/construct? This would avoid a lot of memory allocations.